ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,160 @@
+   * @param crossColoTracker the {@link Histogram} that tracks inter datacenter latencies for this class of requests.
+   * @param pastDueCounter the {@link Counter} that tracks the number of times a request is past due.
+   * @param quantile the quantile cutoff to use for when evaluating requests against the trackers.
+   */
+  AdaptiveOperationTracker(String datacenterName, PartitionId partitionId, boolean crossColoEnabled, int successTarget,
+      int parallelism, Time time, Histogram localColoTracker, Histogram crossColoTracker, Counter pastDueCounter,
+      double quantile) {
+    super(datacenterName, partitionId, crossColoEnabled, successTarget, parallelism, true);
+    this.datacenterName = datacenterName;
+    this.time = time;
+    this.localColoTracker = localColoTracker;
+    this.crossColoTracker = crossColoTracker;
+    this.pastDueCounter = pastDueCounter;
+    this.quantile = quantile;
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    super.onResponse(replicaId, isSuccessFul);
+    long elapsedTime;
+    if (unexpiredRequestSendTimes.containsKey(replicaId)) {
+      elapsedTime = time.milliseconds() - unexpiredRequestSendTimes.remove(replicaId);
+    } else {
+      elapsedTime = time.milliseconds() - expiredRequestSendTimes.remove(replicaId);
+    }
+    getTracker(replicaId).update(elapsedTime);
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();
+    return otIterator;
+  }
+
+  /**
+   * Gets the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that {@code replicaId}
+   * belongs to.
+   * @param replicaId the {@link ReplicaId} whose request latency is going to be tracked.
+   * @return the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that
+   * {@code replicaId} belongs to.
+   */
+  private Histogram getTracker(ReplicaId replicaId) {
+    if (replicaId.getDataNodeId().getDatacenterName().equals(datacenterName)) {
+      return localColoTracker;
+    }
+    return crossColoTracker;
+  }
+
+  /**
+   * An iterator to fetch replicas to send requests to. Respects parallelism and discounts requests that have been
+   * outstanding for a certain amount of time from parallelism.
+   */
+  private class OpTrackerIterator implements Iterator<ReplicaId> {
+
+    @Override
+    public boolean hasNext() {
+      return replicaIterator.hasNext() && (inflightCount < parallelism || isOldestRequestPastDue());
+    }
+
+    @Override
+    public void remove() {
+      replicaIterator.remove();
+      if (inflightCount >= parallelism) {
+        // we are here because oldest request is past due
+        Map.Entry<ReplicaId, Long> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
```

Comment:
Yes, since requests are removed if they are past due and another replica is acknowledged via replicaIterator.remove().